### ISSUE TITLE
Upgrade to SDL2

### ISSUE THREAD

**Issue by M374LX** [LMMS/lmms#2373]
Does everyone agree we should upgrade the SDL audio device to SDL2?

**Comment by tresf** [comment]
Some references... 

http://www.gamedev.net/topic/661386-sdl-vs-sdl2/
https://wiki.libsdl.org/MigrationGuide

Not my wheelhouse, so I'll stay out of the conversation for now. :+1:

**Comment by softrabbit** [comment]
> Not my wheelhouse, so I'll stay out of the conversation for now.

Not mine either, but I just had to take a look, and: 

> The good news for audio is that, with one exception, it's entirely backwards compatible with 1.2. If you want the new features, they're available to you, but you'll probably just compile and run without them. 
> That one really important exception: The audio callback does NOT start with a fully initialized buffer anymore. You must fully write to the buffer in all cases. If you don't have enough audio, your callback should write silence. If you fail to do this, you'll hear repeated audio, or maybe audio corruption. If you want to restore the old behavior of unconditionally initializing the buffer, just put an SDL_memset(stream, 0, len) at the start of your callback. 

So it could be as easy as just telling ```CMake``` to prefer SDL 2 over SDL 1.2, updating any documentation mentioning SDL, and informing packagers?

**Comment by musikBear** [comment]
perhaps allow both? In case some hw dont support sdl2..
User then would have a level of decision.. Kind of a sure contra sorry situation.

**Comment by M374LX** [comment]
According to http://stackoverflow.com/questions/17496316/what-are-the-actual-sdl2-hardware-requirements:
> The minimum system requirements will depend alot more on the application that you are writing than what SDL2 does.

So, SDL2 not supporting older hardware does not seem to be a problem.

**Comment by speedy-software-ch** [comment]
Hey people. I've tried to change the build settings to SDL2 and haven't been able to realize any problems , though i'm not sure it didn't compile without SDL1.2. I uninstalled SDL1.2 via apt-get under Ubuntu. Is that possible?

**Comment by zonkmachine** [comment]
> I've tried to change the build settings to SDL2 and haven't been able to realize any problems , though i'm not sure it didn't compile without SDL1.2. I uninstalled SDL1.2 via apt-get under Ubuntu. Is that possible?

I tried my hands at this but failed miserably. I don't know how to verify it other than uninstalling 1.2 like you did and see if it runs. If you want feedback on this I think you should [submit a Pull Request](https://github.com/LMMS/lmms/wiki/Submitting-a-patch) against the master branch.

**Comment by michaelgregorius** [comment]
> Hey people. I've tried to change the build settings to SDL2 and haven't been able to realize any problems , though i'm not sure it didn't compile without SDL1.2. I uninstalled SDL1.2 via apt-get under Ubuntu. Is that possible?

@speedy-software-ch, if you are compiling under Linux you use the `ldd` command to check which libraries are linked. Pipe to `grep` to check if SDL is somewhere in the results. Execute the following command in the directory where the `lmms` executable was build:
```
ldd lmms | grep SDL
```
In my case the result is that `libSDL-1.2` is linked.
```
        libSDL-1.2.so.0 => /usr/lib/libSDL-1.2.so.0 (0x...)
```

**Comment by speedy-software-ch** [comment]
After I've changed the settings in my environment, build settings and linked *.h files, i got this as a result to the above command.
ldd lmms | grep SDL
	libSDL2-2.0.so.0 => /usr/lib/x86_64-linux-gnu/libSDL2-2.0.so.0 (0x00007fe93d52c000)
So i think it works with SDL 2.0. Do you want me to commit this changes back to the github repository?

**Comment by michaelgregorius** [comment]
Looks good, @speedy-software-ch! The "normal" way to integrate changes is as follows:

1. Create a separate local feature branch and commit your changes there.
2. Push this local branch to your remote LMMS repository on Github.
3. Create a pull request for that branch.

That way other developers can review and test them before they are merged into master.

Is https://github.com/speedy-software-ch/lmms/tree/SDL1.2-to-SDL2 the branch that contains your changes? If so you will only need to issue a pull request for that branch. However, the branch seems to be based on the 1.2 (LMMS version not SDL) branch and that might need some discussion. I think changes usually go into master first and they are then merged selectively into 1.2.

**Comment by tresf** [comment]
Agreed, `master` is the better branch for this.  I've left some comments on the PR [here](https://github.com/LMMS/lmms/pull/3509#issuecomment-296811571).

**Comment by tresf** [comment]
Closed via a1480fe.
